proof/refine/ARM/Syscall_R.thy

```diff
@@ -2677,6 +2593,400 @@ proof -
+   apply (clarsimp simp: active_sc_at'_def obj_at'_def projectKOs cur_tcb'_def pred_tcb_at_def
+                         is_sc_obj obj_at_def)
+   apply (clarsimp simp: pred_map_def isScActive_def)
+   apply (rule_tac x="cur_sc s" in exI)
+   apply (clarsimp simp: opt_map_red dest!: state_relationD)
+   apply (frule_tac x="ksCurThread s'" in pspace_relation_absD, simp)
+   apply (fastforce simp: other_obj_relation_def tcb_relation_def)
+  apply (rule_tac Q="\<lambda>s'. pred_map (\<lambda>tcb. \<not> tcbInReleaseQueue tcb) (tcbs_of' s') (ksCurThread s')"
+         in corres_cross_add_guard)
+   apply (clarsimp, frule tcb_at_invs)
+   apply (fastforce simp: not_in_release_q_def release_queue_relation_def pred_map_def opt_map_red obj_at'_def
+                          invs'_def valid_pspace'_def projectKOs valid_release_queue'_def cur_tcb'_def
+                   dest!: state_relationD)
+  apply (rule corres_rel_imp)
+   apply (rule corres_guard_imp)
+     apply (rule corres_split_liftEE[OF getCapReg_corres_gen])
+       apply (rule corres_splitEE[OF _ handleInvocation_corres])
+           apply (rule_tac P="(einvs and ct_active and scheduler_act_sane and current_time_bounded 2
+                              and ct_not_queued and ct_not_in_release_q)"
+                      and P'="(invs')"
+                  in corres_inst)
+           apply (rule corres_stateAssertE_add_assertion)
+            apply simp
+            apply (rule corres_stateAssertE_add_assertion)
+             apply (rule corres_stateAssertE_add_assertion)
+              apply simp
+              apply (clarsimp simp: sch_act_sane_asrt_def ct_not_ksQ_asrt_def ct_active'_asrt_def)
+              apply (rule corres_guard_imp)
+                apply (rule handleRecv_isBlocking_corres)
+               apply simp
+              apply simp
+             apply (fastforce simp: ct_active'_asrt_def invs_psp_aligned invs_distinct
+                             dest!: ct_active_cross)
+            apply (clarsimp simp: ct_not_ksQ_asrt_def not_queued_2_def ready_queues_relation_def
+                           dest!: state_relationD)
+           apply (clarsimp simp: sch_act_sane_asrt_def scheduler_act_not_def sch_act_sane_def sched_act_relation_def
+                           dest!: state_relationD)
+           apply (case_tac "scheduler_action s"; simp)
+          apply simp
+         apply simp
+        apply ((wpsimp wp: handle_invocation_valid_sched
+              | strengthen current_time_bounded_strengthen[where k=5])+)[1]
+       apply wpsimp
+      apply wpsimp
+     apply wpsimp
+    apply (clarsimp simp: invs_def valid_state_def valid_pspace_def active_from_running
+                          schedulable_def2 released_sc_tcb_at_def)
+    apply (clarsimp simp: ct_in_state_def st_tcb_weakenE)
+   apply (clarsimp simp: active_from_running')
+  apply simp
+  done
+
+abbreviation (input)
+  "a_pre \<equiv> (einvs and ct_running and
```

Comment:
This is almost the same as `mcs_invs`. Can they be consolidated or reused somehow?

Reply:
These abbreviations (`pre_a` and `pre_c`) are only meant to be used very locally in the two helper lemmas that follow, and the purpose is to ensure that they should have same preconditions. I wouldn’t have needed to introduce them if I could unify these two helper lemmas that are almost identical but in different monads. It would be nice if we can abstract away the structure somehow to avoid duplication, but our current corres framework doesn’t seem to allow that.